### ISSUE TITLE
Yolov8 OBB Task support (Oriented Bounding Box - Object Detection)

### ISSUE THREAD

**Issue by JustasBart** [ultralytics/ultralytics#624]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hi guys,

I'm just wondering of what could be the likelihood of discussing the OBB support for yolov8 sometime in the future?

I know that based on the 'Ultralytics Live Session of January 24th' there are certain priorities that will have to be taken care of first in order to bring the maturity of yolov8 to the point of yolov5 and that later there might be talk about implementing the Pose estimation etc...

But I've just been wondering if OBB support might ever make it on the table for this project?

I'm aware that there already exist certain options for training OBB models and even a specialised version of yolov5 that implements this.
However, none are quite as nicely maintained and as quick and easy to implement as the Ultralytics based project repositories and I would therefore love to see an OBB option with yolov8.

If I knew how I would even go as far as to implement it myself, but what do you guys think? Is this doable? Likely?

Thanks!

### Use case

In addition to the obvious scenarios (Such as satellite imagery of vehicles) I could totally see how this would work really well with things like electronics (Resistors/fuses/wire jumps/etc...) it should in theory help a lot when it comes to having a small enough dataset and having a good bit of overlap (Perhaps from parts being close together).

It would mainly benefit any objects that are stretched out in length where the bounding box is good if they are either vertical (|) or horizontal (---) but it becomes messy if they turn on say 45 degrees (\\) whereas with OBB it wouldn't matter.

Which could include things like:
- All sorts of Medical and other types equipment.
- Electricity poles (For things like self driving cars etc...).
- Cutlery (Forks, spoons, knives, etc...).
- Agriculture (For example for things like carrots on a conveyor belt).
- Automated electronics placement/inspection etc...

I think that there are endless examples of where this would be really useful to have as an option.

![Standard](https://user-images.githubusercontent.com/40023722/214549351-47f59808-748c-4980-ba76-9821461f5207.jpg)
versus
![OBB](https://user-images.githubusercontent.com/40023722/214549376-c00a9f16-934b-4713-a00e-d64e86b66408.jpg)

Note the overlap of the components in each case...

### Real-World-Examples

Consider this case case:
![SIPResistors_Good_Sample](https://user-images.githubusercontent.com/40023722/215073783-ae4e9dbc-a4fa-411a-ac20-06568710593e.png)

Or this case:
![Capacitors](https://user-images.githubusercontent.com/40023722/215073819-0f5bc95a-6b8b-467f-82eb-a9717f1fbf6c.png)

Both of these cases will work perfectly fine with the current detection boxes.

However, now consider if the yellow (SIP_Resistor) parts were on an angle like so:
![SIPResistors2](https://user-images.githubusercontent.com/40023722/215074023-e5f4b6cb-974b-4d49-a2ad-dd97b74c86fb.png)
![SIPResistors](https://user-images.githubusercontent.com/40023722/215074026-4f86de6a-e905-4098-807e-5817d1200dad.png)

Or consider this case right here (Where the shape of the object is a square, but the object can be on an angle):
![Terminals](https://user-images.githubusercontent.com/40023722/215074217-2254014a-9e2d-41a8-a6a3-e77ca4c15850.png)

Or let's consider that we don't have that much data to begin with, but that we have components that are next to each other like so:
![LEDs](https://user-images.githubusercontent.com/40023722/215074935-0dfdeaf9-eaeb-4d37-a9b7-7ea3c3cfc1d6.png)

### Real-World-Examples Using OBB

Now let's consider these tricky samples with OBB:
![SIPResistors_OBB2](https://user-images.githubusercontent.com/40023722/215095236-bb8db853-c2f7-4844-8e28-0480ac9cd0d8.png)
![SIPResistors_OBB](https://user-images.githubusercontent.com/40023722/215095365-d7df6823-3633-4c89-a5f7-eee6e36083ff.png)
![Terminals_OBB](https://user-images.githubusercontent.com/40023722/215095495-84572e5e-c8fb-401a-9de5-388d4663d484.png)
![LEDs_OBB](https://user-images.githubusercontent.com/40023722/215095572-e37db3e2-6653-4e56-aaee-e6d8068b54e0.png)

I've ashed ChatGPT to build me a basic OBB labelling tool (Does not produce the actual labels) [OBB Labeling tool Python code](https://pastebin.com/RU8PR6aJ) and ironically for this dataset where the component sizes are more or less identical it was nearly easier and faster to label than the traditional labelling tools/approaches...

### Additional

This is definitely a low priority in comparison to what has to happen first, but I still firmly believe that this is worth looking into for the longer-term cycle of yolov8.

In the mean time I just wanted to bring it up for discussion :rocket: 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by AyushExel** [comment]
@JustasBart This is one of my favorite feature requests! Thanks for the detailed write-up. It perfectly explains the use case.
Yes, we might be able to support that but I can't provide a date. It depends a lot on how complex it is going to be. 
If you'd like to work on this, I'd be happy to assist you. Please contact me on twitter / LinkedIn and let's setup a call to discuss this.

**Comment by mikel-brostrom** [comment]
You would have to add a component to the loss function as suggested in this paper: https://arxiv.org/pdf/1803.06199.pdf. Training could be done on KITTI 3D object detection by projecting the point clouds into the ground plane and using their associated 3D bboxes projected into the ground plane.

**Comment by jeck00119** [comment]
I found a list of projects that implemented OOB, maybe it will help someone.
[https://github.com/sjinzh/awesome-yolo-object-detection#oriented-object-detection](url)

**Comment by JustasBart** [comment]
@AyushExel I've added some real-world examples as an additional reference :rocket:

**Comment by regainOWO** [comment]
@JustasBart highly recommend [hukaixuan19970627/yolov5_obb](https://github.com/hukaixuan19970627/yolov5_obb) . it use increase 180 dim for no in Detect to class bounding box angle, not a regression method.
also the repo [open-mmlab/mmrotate](https://github.com/open-mmlab/mmrotate) is very usefull.

**Comment by a15097916856** [comment]
I think the target detection of rotating frame is really a valuable improvement direction. I hope YOLOV8 can support OBB

**Comment by haimat** [comment]
Agreed, would have had quite some use cases for OBB in the past.
Definitely a +1 from me :+1:

**Comment by yoloyash** [comment]
@AyushExel Looking into this right now. Will try to implement it first and if I have some decent results maybe we can get on a call and discuss it further. Yeah? Would you guys be interested in merging a PR relating to OBB?

**Comment by AyushExel** [comment]
@yoloyash yes! Do let me know if you get some nice results

**Comment by glenn-jocher** [comment]
@yoloyash yes definitely. I'm not sure if we'd want to create an entirely new task for this (probably), or call this an extension of detect with one more rotation parameter (also an option). 

What's the main reference OBB dataset, i.e. the COCO of OBB?

**Comment by yoloyash** [comment]
@glenn-jocher We can use the DOTA dataset for training and evaluation for our reference

**Comment by glenn-jocher** [comment]
@yoloyash That sounds like a great starting point. Let's work towards making OBB support in YOLOv8 a reality, and we can share our progress with the community along the way. Thank you for your initiative on this!

**Comment by glenn-jocher** [comment]
@yoloyash linked to your PR #1958 :)

**Comment by ZeonlungPun** [comment]
what is the definition of  angle in yolov8?
![image](https://github.com/ultralytics/ultralytics/assets/122013127/7d7a83c1-1800-4371-9b8c-a39ef401c9a5)

**Comment by glenn-jocher** [comment]
@ZeonlungPun the angle in YOLOv8 refers to the orientation of the object with respect to the horizontal axis. It's measured in degrees, with positive values indicating a counter-clockwise rotation and negative values for clockwise rotation. 🔄

**Comment by ZeonlungPun** [comment]
> @ZeonlungPun the angle in YOLOv8 refers to the orientation of the object with respect to the horizontal axis. It's measured in degrees, with positive values indicating a counter-clockwise rotation and negative values for clockwise rotation. 🔄

thank you for your reply. does it mean that the range of angle is [-pi/2,pi/2] ? but in the source code (ultralytics/nn/modules
/head.py) , i see this code in forward process :  angle = (angle.sigmoid() - 0.25) * math.pi  # [-pi/4, 3pi/4]  ; it means that the range of angle is  [-pi/4, 3pi/4] , which makes me confusued.

**Comment by glenn-jocher** [comment]
@ZeonlungPun The angle output from the model is indeed transformed from the sigmoid output to a range of `[-pi/4, 3pi/4]` in the code. This range is used to ensure that the model can predict angles for all possible orientations of objects. The transformation from the sigmoid function to this range allows the model to effectively learn and predict the angle of orientation. 📐

**Comment by bwsyh** [comment]
> @ZeonlungPun The angle output from the model is indeed transformed from the sigmoid output to a range of `[-pi/4, 3pi/4]` in the code. This range is used to ensure that the model can predict angles for all possible orientations of objects. The transformation from the sigmoid function to this range allows the model to effectively learn and predict the angle of orientation. 📐

I use yolov8n-obb.pt to convert to ONNX model, output [1,20,21504], is the Angle at the end of the output 20 data?
It returns a value of 1.00988257, how do I get the Angle of rotation at the center of the rectangle?
My friend is also confused about this, or where can I check the relevant help documentation?
Looking forward to your reply!

**Comment by bwsyh** [comment]
@glenn-jocher 
I use yolov8n-obb.pt to convert to ONNX model, output [1,20,21504], is the Angle at the end of the output 20 data?
It returns a value of 1.00988257, how do I get the Angle of rotation at the center of the rectangle?
My friend is also confused about this, or where can I check the relevant help documentation?
Looking forward to your reply!

**Comment by glenn-jocher** [comment]
@bwsyh the output you're referring to likely includes the angle information as part of the 20 data points for each detection. The value you see needs to be interpreted within the context of the model's output format. For the specific transformation and interpretation of the angle value, please refer to the post-processing steps detailed in our documentation. If you need further clarification, the documentation on the Predict mode will be particularly helpful. 📚

**Comment by karthikyerram** [comment]
@glenn-jocher is yolov8 obb deployment supported at ultralytics hub? can we do the inference yet?

**Comment by glenn-jocher** [comment]
@karthikyerram Yes, YOLOv8 OBB deployment is supported on the Ultralytics Hub. You can perform inference using the available models. For guidance on how to proceed with inference, please refer to the Predict mode documentation. 🚀

**Comment by ZeonlungPun** [comment]
What are the main mathematical principles or ideas for implementing this function?

def dist2rbox(pred_dist, pred_angle, anchor_points, dim=-1):
    """
    Decode predicted object bounding box coordinates from anchor points and distribution.

    Args:
        pred_dist (torch.Tensor): Predicted rotated distance, (bs, h*w, 4).
        pred_angle (torch.Tensor): Predicted angle, (bs, h*w, 1).
        anchor_points (torch.Tensor): Anchor points, (h*w, 2).
    Returns:
        (torch.Tensor): Predicted rotated bounding boxes, (bs, h*w, 4).
    """
    lt, rb = pred_dist.split(2, dim=dim)
    cos, sin = torch.cos(pred_angle), torch.sin(pred_angle)
    # (bs, h*w, 1)
    xf, yf = ((rb - lt) / 2).split(1, dim=dim)
    x, y = xf * cos - yf * sin, xf * sin + yf * cos
    xy = torch.cat([x, y], dim=dim) + anchor_points
    return torch.cat([xy, lt + rb], dim=dim)

**Comment by glenn-jocher** [comment]
@ZeonlungPun hi there,

Thanks for bringing up the topic of Oriented Bounding Box (OBB) support for YOLOv8. It's great to see such enthusiasm and detailed use cases from the community! 🚀

While OBB is not currently supported in YOLOv8, we're always considering new features that can enhance the capabilities of our models. The examples you've provided clearly illustrate the potential benefits of OBB for certain types of object detection tasks.

Implementing OBB would indeed be a significant undertaking, and as you mentioned, there are priorities to ensure YOLOv8 reaches maturity. However, we're open to exploring this feature in the future, and community interest like yours is a key factor in our decision-making process.

If you or anyone else is interested in contributing to this effort, we welcome collaboration and contributions. While we can't promise immediate action on this feature, we'll certainly keep it on our radar as we continue to develop YOLOv8.

Thanks again for your input, and let's keep the conversation going! 🤝

**Comment by dimidagd** [comment]
> @ZeonlungPun hi there,
> 
> Thanks for bringing up the topic of Oriented Bounding Box (OBB) support for YOLOv8. It's great to see such enthusiasm and detailed use cases from the community! 🚀
> 
> While OBB is not currently supported in YOLOv8, we're always considering new features that can enhance the capabilities of our models. The examples you've provided clearly illustrate the potential benefits of OBB for certain types of object detection tasks.
> 
> Implementing OBB would indeed be a significant undertaking, and as you mentioned, there are priorities to ensure YOLOv8 reaches maturity. However, we're open to exploring this feature in the future, and community interest like yours is a key factor in our decision-making process.
> 
> If you or anyone else is interested in contributing to this effort, we welcome collaboration and contributions. While we can't promise immediate action on this feature, we'll certainly keep it on our radar as we continue to develop YOLOv8.
> 
> Thanks again for your input, and let's keep the conversation going! 🤝

If OBB is not supported, then what about https://github.com/ultralytics/ultralytics/pull/4499 ?

**Comment by JustasBart** [comment]
@dimidagd Hi, that was just humorous Glen AI having a bit of fun with us there, the OBB is indeed supported for some time now! Really glad to see this coming full circle and finally being a part of yolov8 🚀 

All the best!

**Comment by VenkateshSoni** [comment]
How can I get the bounding boxes of the OBB model?
`
results = model.predict(frame, save=True, imgsz=640, conf=0.5)

    for r in results:
        boxes = r.boxes
        for box in boxes:
            bbox = box.xyxy[0]  # get box coordinates in (left, top, right, bottom) format
            print(bbox)
            npbbox = bbox[0:4].numpy().astype(int)  # Convert to integers
            print(npbbox)
`
`
    for r in results:
        boxes = r.boxes
---> for box in boxes:
       bbox = box.xyxy[0]
       print(bbox)

TypeError: 'NoneType' object is not iterable
`

**Comment by glenn-jocher** [comment]
@VenkateshSoni thank you for your detailed feature request regarding Oriented Bounding Box (OBB) support for YOLOv8. Your insights into potential use cases and the benefits of OBB for certain types of objects are well-articulated and highlight the value that this feature could bring to a variety of applications.

As you've mentioned, there are existing solutions for OBB detection, including specialized versions of YOLOv5. However, the ease of use and maintenance quality of Ultralytics projects make them a preferred choice for many users.

While I cannot speak to the likelihood of OBB support being added to YOLOv8 in the near future, I can say that the Ultralytics team is always interested in exploring new features that could enhance the capabilities of our models. The implementation of OBB would indeed be a significant undertaking, requiring modifications to the model architecture, training procedures, and post-processing steps.

Given the current priorities and roadmap for YOLOv8, it's challenging to provide a definitive answer on whether and when OBB support might be officially integrated. However, the team is open to community contributions, and if you or anyone else in the community is interested in working on this feature, we would be happy to discuss potential collaboration.

In the meantime, I will bring this feature request to the attention of the Ultralytics team for further consideration. We appreciate your enthusiasm and the examples you've provided, which clearly demonstrate the potential impact of OBB support.

If you have any further questions or ideas, please feel free to continue the discussion here or on our GitHub repository. Your feedback is invaluable as we continue to develop and improve YOLOv8.

**Comment by rbgreenway** [comment]
@glenn-jocher I've been very interested in OBB for object detection for quite some time so I'm excited to see it supported now.  I would like to ask the community if there is any sense whether this approach improves the detector?  I've long wondered if given the ability to OBB annotate objects with large aspect ratios (like trombones or baseball bats, for example) allows for far less background contamination in the bounding box...particularly when the object is oriented away from horizontal or vertical. If anyone has any thoughts or experience with this, I'd love to hear it.  Thanks for a great set of tools and a well-supported, active community!

**Comment by molnarszilard** [comment]
> What are the main mathematical principles or ideas for implementing this function?
> 
> def dist2rbox(pred_dist, pred_angle, anchor_points, dim=-1): """ Decode predicted object bounding box coordinates from anchor points and distribution.
> 
> ```
> Args:
>     pred_dist (torch.Tensor): Predicted rotated distance, (bs, h*w, 4).
>     pred_angle (torch.Tensor): Predicted angle, (bs, h*w, 1).
>     anchor_points (torch.Tensor): Anchor points, (h*w, 2).
> Returns:
>     (torch.Tensor): Predicted rotated bounding boxes, (bs, h*w, 4).
> """
> lt, rb = pred_dist.split(2, dim=dim)
> cos, sin = torch.cos(pred_angle), torch.sin(pred_angle)
> # (bs, h*w, 1)
> xf, yf = ((rb - lt) / 2).split(1, dim=dim)
> x, y = xf * cos - yf * sin, xf * sin + yf * cos
> xy = torch.cat([x, y], dim=dim) + anchor_points
> return torch.cat([xy, lt + rb], dim=dim)
> ```

This is still an unanswered question, which interests me as well. Could you explain this function in detail without using Glen AI? It would be extremely useful. 

Or better yet: do you have scientific articles or documentation where the functions are explained (the [Docs](https://docs.ultralytics.com) is quite vague, and sometimes old)? I can imagine that you have a lot of things to do, but detailed documentation would free you in terms of dealing less with issues like this one.

**Comment by glenn-jocher** [comment]
@molnarszilard hey there! 🌟

Absolutely, I'd be happy to shed some light on the `dist2rbox` function. This function is designed to decode predicted bounding box coordinates from given anchor points, predicted distances, and angles to their rotated counterparts. Here's a quick breakdown:

1. **Splitting Predicted Distances**: The function starts by splitting the predicted distances (`pred_dist`) into two parts - left-top (`lt`) and right-bottom (`rb`) distances. This is done to separate the distances on different sides of the anchor points.

2. **Calculating Cosine and Sine**: It calculates the cosine and sine of the predicted angles (`pred_angle`). These values are used to rotate the bounding box coordinates.

3. **Calculating Adjustments**: The function calculates adjustments (`xf`, `yf`) for the x and y coordinates. This is essentially finding the half-point between `lt` and `rb` to determine how far to move in each direction.

4. **Applying Rotation**: The adjustments are then rotated by applying the cosine and sine values calculated earlier. This step effectively rotates the bounding box.

5. **Adjusting Anchor Points**: The rotated adjustments are added to the anchor points to position the bounding box correctly in the image space.

6. **Concatenating Results**: Finally, it concatenates the adjusted anchor points with the total distances (`lt + rb`) to form the complete rotated bounding box coordinates.

Regarding documentation and scientific articles, while specific articles on this exact function might be scarce, the principles it employs are rooted in basic trigonometry and object detection methodologies. For a deeper dive, I'd recommend looking into materials on geometric transformations in computer vision, as well as the YOLO (You Only Look Once) series of papers for foundational knowledge on object detection.

I hope this helps clarify things a bit! If you have more questions or need further assistance, feel free to reach out. Happy coding! 🚀

**Comment by molnarszilard** [comment]
Hi!
At the first point ```pred_dist``` is a distance or distribution? If it is a distance, than it is a distance between what exactly?

**Comment by glenn-jocher** [comment]
@molnarszilard hi there! 👋

Great question! In the context of the `dist2rbox` function, `pred_dist` refers to the predicted distances from the center of the anchor point to the edges of the oriented bounding box (OBB). Essentially, it represents the distances to the left-top and right-bottom edges of the bounding box from its center. This helps in accurately positioning and sizing the OBB around the detected object.

Hope this clears things up! If you have any more questions, feel free to ask. Happy coding! 🚀

**Comment by molnarszilard** [comment]
Let's see if I can phrase it in my words...
I have the anchor points, which are the intersections of a grid on the image. (By default the image is downscaled according to the last layers of the [model](https://github.com/ultralytics/ultralytics/issues/189#issue-1527158137). There in the Head, I have 3 layers (e.g. starting from an image of 640x640 the last layer will have the sizes: 80x80, 40x40, 20x20, and the points are situated on every half increment: 0.5,1.5,...,79.5. The total number of anchor points is 8400). 

The model then predicts the same number of feature vectors (each feature vector containing number_of_classes+4*16 values - 16 is the [Distribution Focal Loss (DFL)](https://ieeexplore.ieee.org/document/9792391) channels), therefore each feature vector can be linked to a specific anchor point.

In ```pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3).matmul(self.proj.type(pred_dist.dtype))```, 
To obtain the ```pred_dist``` tensor, I take the 4*16 part of the feature vector, apply a [softmax](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html) on the last dimension (softmax rescales the elements over a vector in such a way, that the sum of the elements is one, while each element is in the range of [0,1]). Then a [matmul](https://pytorch.org/docs/stable/generated/torch.matmul.html) operation is performed, where the ```input``` is the output of the softmax and the ```other``` is a vector=[0,1,2,...,15]. I am not sure what this does exactly (My intuition was that it multiplies the 16 values by elements of the vector, then sums them up, therefore the 16 values become 1 -- out_softmax=[0.21,0.023,...,0.005], then matmul=[0.21*0+0.023*1+...+0.005*15]? However, then the first element will always be 0... So I think this part is unclear, even when reading the documentation of the function).

Then in the ```dist2rbox``` function gets the ```pred_dist``` tensor as (bs,number_of_anchors,4). Then for each anchor point, we consider the current anchor point to be the origin of the current coordinate system. Then the ```pred_dist``` is split into two: ```lt``` and ```lb```. In the current coordinate system, the top-left point will in fact have the coordinates (-left,-top), while the bottom-right point will have the coordinates (+right, +bottom).

```x, y = xf * cos - yf * sin, xf * sin + yf * cos```
```xf``` and ```yf``` describe the center between the top-left and bottom-right points. Then this center point is rotated along the origin according to the ```pred_angle```. This rotated point will have the coordinates ```x``` and ```y```.

To move this rotated center point to the coordinate system of the image, we add to them the coordinates of the current anchor point. This coordinate system-adjusted center point is considered as the center point of the predicted bounding box,  while the width and height of the bounding boxes come from the ```lt``` and ```lb``` values.

This is how I understood it. Do you have something to add? (Maybe this description could be useful for other people too, especially if in some forms it would appear in formal documentation or at least in the code).

**Comment by glenn-jocher** [comment]
@molnarszilard hey there! 🌟

You've done a fantastic job breaking down the process, and your understanding is pretty spot on! The way you've described the transformation from the feature vectors to the final oriented bounding box (OBB) coordinates is quite accurate. 

Just to add a bit on the `matmul` part you mentioned: your intuition is correct. The operation essentially weights and sums up the softmax outputs using the vector [0,1,2,...,15]. This step helps in decoding the distances from the anchor points to the edges of the bounding box. The reason the first element isn't always 0 is due to the softmax operation applied before, which ensures that the distribution of the 16 channels contributes to the final distance value in a meaningful way.

Your detailed walkthrough is indeed valuable and could serve as a great starting point for more formal documentation or code comments. It's always beneficial to have different perspectives on understanding these operations, especially when it comes to complex transformations like these.

Thanks for sharing your insights, and I'm sure many will find your explanation helpful! Keep up the great work. 🚀

**Comment by SFSimDev** [comment]
I'm toying with the new OBB implementation to replace some existing code and wanted to say thanks for implementing it.

I have a couple of queries:

1. The pre-trained yolov8x-obb model works pretty well at detecting images in a GeoTIFF, BUT how would I get it to preserve the coordinates from the GeoTIFF Coordinate Reference System in the output (is this a feature or one that could be implemented?)
2. How can I output the results in GeoJSON format (or JSON for that matter)? I've tried print(results[0].tojson()) but without success.

**Comment by glenn-jocher** [comment]
Hey there! 🌟

Glad to hear you're exploring the new OBB feature in YOLOv8! Let's dive right into your queries:

1. **Preserving GeoTIFF Coordinates**: The current version might not directly support extracting and preserving GeoTIFF coordinates in the output. However, this is a valuable suggestion! While we work on potentially including it as a feature, you could implement a workaround by utilizing GDAL or rasterio to read the GeoTIFF's CRS (Coordinate Reference System) and manually append this info to your detection outputs.

2. **Output in GeoJSON Format**: To output the results in GeoJSON or JSON formats, ensuring you're using the latest version of Ultralytics YOLOv8 is key. If `results[0].tojson()` isn't working as expected, you might try serializing the results manually for now. Here’s a quick way to get your detections into JSON:
   ```python
   import json
   results_json = [result.pandas().xyxy[0].to_dict(orient="records") for result in results]
   json_str = json.dumps(results_json)
   print(json_str)
   ```
   For GeoJSON, you'd need to embed this JSON inside a GeoJSON structure, taking care to insert the appropriate geographic coordinates.

Hope this helps, and keep the feedback coming! It's great to see what you're building. 🚀

**Comment by glenn-jocher** [comment]
Hey there! 😊

Absolutely, it sounds like a great idea! Feel free to raise it as a feature request on the GitHub. Your suggestions help us improve and add valuable capabilities to the project. Thanks for thinking along! If you've got a code example or specific details in mind, including those in your request would be super helpful too. Looking forward to seeing your contributions! 🚀

**Comment by Petros626** [comment]
note for me.

**Comment by Petros626** [comment]
> What are the main mathematical principles or ideas for implementing this function?
> 
> def dist2rbox(pred_dist, pred_angle, anchor_points, dim=-1): """ Decode predicted object bounding box coordinates from anchor points and distribution.
> 
> ```
> Args:
>     pred_dist (torch.Tensor): Predicted rotated distance, (bs, h*w, 4).
>     pred_angle (torch.Tensor): Predicted angle, (bs, h*w, 1).
>     anchor_points (torch.Tensor): Anchor points, (h*w, 2).
> Returns:
>     (torch.Tensor): Predicted rotated bounding boxes, (bs, h*w, 4).
> """
> lt, rb = pred_dist.split(2, dim=dim)
> cos, sin = torch.cos(pred_angle), torch.sin(pred_angle)
> # (bs, h*w, 1)
> xf, yf = ((rb - lt) / 2).split(1, dim=dim)
> x, y = xf * cos - yf * sin, xf * sin + yf * cos
> xy = torch.cat([x, y], dim=dim) + anchor_points
> return torch.cat([xy, lt + rb], dim=dim)
> ```

@ZeonlungPun sorry, but the image you're referring to is the following:
![image](https://github.com/user-attachments/assets/fc89c962-844d-447b-bcf8-bd0089df994c)


so when YOLOv8 OBB uses intern `cv2.minAreaRect` and you got OpenCV>=4.5.1positive values are in CW and negative values in CCW. The same for the output of `[-pi/4...3pi/4]` or not?

**Comment by Kang815** [comment]
realtime detection is performed using yolov8-obb, anchor boxes are created and log files are generated in the order of milliseconds, reliability, and classes  but I'm fail why?

**Comment by glenn-jocher** [comment]
@Kang815 it seems like your issue might be related to your dataset, model configuration, or input data. Please ensure your dataset is correctly formatted for OBB (see [OBB Dataset Guide](https://docs.ultralytics.com/datasets/obb/)) and verify your model inputs match the expected format. If the problem persists, share more details like error messages or logs for further assistance.
